### PR TITLE
fix(swap): decode the offer TLV fields solverd emits

### PR DESCRIPTION
**Pull request by Kukks**
`decodeOffer` throws `unknown TLV type: 0xc` on any offer `solverd` publishes.

## The gap

solverd's maker helper builds the maker's unilateral exit closure **by default** —
it is opt-*out* (`pkg/swap/contract/maker.go:97`, `if !params.NoExit`), using the
server's `UnilateralExitDelay` — and emits it as TLV tag `0x0c`. Our `FIELDS`
table knows tags `0x01`, `0x02`, `0x03`, `0x05`, `0x07`, `0x08` and `0x0b` only,
and the decoder is strict by design, so the entire payload is refused.

Tags `0x09`/`0x0a` (the partial-fill ratio) are reserved wire space today —
solverd round-trips them but never sets them — and fail the same way the moment
anything does.

The effect is that a solverd-made offer is undecodable by this SDK. It never
reaches `offerVtxoScript`, `classifySpend`, `fillOffer` (#785, which calls
`decodeOffer` directly), or any acceptance gate a taker puts in front of them.

## The fix

- **Decode and encode `0x09` ratioNum, `0x0a` ratioDen and `0x0c` exitTimelock**,
  in the canonical record order § 2.1 of the swap protocol requires.
- **Build the exit closure into the taproot tree** when `exitDelay` is present: a
  `CSVMultisigClosure` of the maker alone, appended as the third leaf. This has
  to land with the decoding rather than after it — reading the tag without
  building the leaf derives a different address, and every such offer would then
  fail the § 5.1 consistency check instead of decoding.
- **`createOffer` now builds the exit closure by default**, at the server's own
  `unilateralExitDelay`, matching solverd. Without it `cancel` is the only route
  out of an offer, and `cancel` is a 2-of-2 with the server: a server that will
  not co-sign leaves the deposit stuck at the swap address until the VTXO expires
  and the operator sweeps it — and an offer has no expiry of its own, so that
  exposure has no end. `noExit: true` opts out; an explicit `exitDelay`
  overrides.

The delay is derived under the same sub-512-is-blocks threshold arkd's own
closures (`wallet.ts`'s `delayToTimelock`) and solverd's `fetchServerConfig` use.
A server reporting no usable delay is refused rather than published as a CSV of
zero, which would be an offer claiming an exit it does not have.

Ratios are carried, never interpreted — they do not touch the script, which the
`ratio 1/4` vector below pins by deriving the same address as its ratio-less twin.

## What this changes for existing callers

Only **new** offers from `createOffer` get the extra leaf, and their swap address
moves accordingly. A stored `offerHex` carries its own exit field, so it
re-derives to exactly the address it was funded at either way, and `cancelOffer`
and `classifySpend` keep working on offers made before this change.

The exit adds no new race: `cancel` is already **untimelocked**, so the maker can
already spend out from under a fill at any moment. `exit` is strictly more
constrained than a path that already exists.

### Checked against `arkade-os/wallet` (`master`, 179e0b3)

It is the real consumer of this API, so I read it rather than assuming:

- It **imports** `createOffer`/`cancelOffer`/`decodeOffer` from this package and
  holds no copy of the offer construction, the TLV codec or the program JSONs —
  so there is no second implementation to keep in step.
- `createSwap` (`src/providers/assetSwaps.tsx:202`) passes neither `exitDelay`
  nor `noExit`, so it picks up the exit closure with no wallet change. New offers
  derive a different swap address; it persists `offerHex` and passes it back to
  `cancelOffer`, so offers made before the bump still re-derive to the address
  they were funded at.
- Nothing there reads leaves, `tapLeafScript` or `functionByName`, so no code
  assumes a two-leaf tree.
- Its one piece of local swap logic, `isCancelSpend` (`src/lib/swapSpend.ts`),
  is inferential — it asks what the spend moved, not which leaf it took — and
  reports an exit spend as a cancel. That is the same answer `classifySpend` now
  gives, so the wallet's heuristic and this package agree rather than diverging
  on the new leaf.

## Verification

The `solverd reference vectors` block is generated by **solverd's own encoder**,
not by this SDK, so it pins the wire format and the taproot tree against the
reference rather than against ourselves. Each vector asserts that we decode the
payload, re-derive its `swapPkScript` and address, get the right leaf count, and
re-emit byte-identical bytes.

Two of the vectors carry no exit and reproduce the goldens already at the top of
`offer.test.ts` — that is what says the generator ran against the same inputs.

<details>
<summary>Generator (drop in <code>arkade-os/solver</code> at <code>cmd/offervectors/main.go</code>, <code>go run ./cmd/offervectors</code>)</summary>

```go
package main

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"strings"

	arklib "github.com/arkade-os/arkd/pkg/ark-lib"
	"github.com/arkade-os/arkd/pkg/ark-lib/asset"
	"github.com/arkade-os/solver/pkg/swap/contract"
	"github.com/btcsuite/btcd/btcec/v2"
	"github.com/btcsuite/btcd/btcec/v2/schnorr"
)

func xonly(h string) *btcec.PublicKey {
	b, _ := hex.DecodeString(h)
	k, err := schnorr.ParsePubKey(b)
	if err != nil {
		panic(err)
	}
	return k
}

func assetId(h string) *asset.AssetId {
	b, _ := hex.DecodeString(h)
	id, err := asset.NewAssetIdFromBytes(b)
	if err != nil {
		panic(err)
	}
	return id
}

func main() {
	server := xonly("4f355bdcb7cc0af728ef3cceb9615d90684bb5b2ca5f859ab0f0b704075871aa")
	maker := xonly("3c72addb4fdf09af94f0c94d7fe92a386a7e70cf8a1d85916386bb2535c7b1b1")
	emulator := xonly("466d7fcae563e5cb09a0d1870bb580344804617879a14949cf22285f1bae3f27")
	makerPkScript, _ := hex.DecodeString(
		"51203c72addb4fdf09af94f0c94d7fe92a386a7e70cf8a1d85916386bb2535c7b1b1",
	)
	// the same 34-byte id offer.test.ts builds with AssetId.fromString("aa".repeat(32) + "0000")
	testAsset := assetId(strings.Repeat("aa", 33) + "0000")

	blocks := arklib.RelativeLocktime{Type: arklib.LocktimeTypeBlock, Value: 144}
	seconds := arklib.RelativeLocktime{Type: arklib.LocktimeTypeSecond, Value: 51200}
	long := arklib.RelativeLocktime{Type: arklib.LocktimeTypeSecond, Value: 604672}
	short := arklib.RelativeLocktime{Type: arklib.LocktimeTypeBlock, Value: 4032}

	cases := []struct {
		label string
		want  *asset.AssetId
		offer *asset.AssetId
		exit  *arklib.RelativeLocktime
		ratio [2]uint64
	}{
		{label: "wantBtc, no exit", offer: testAsset},
		{label: "wantBtc, exit blocks 144", offer: testAsset, exit: &blocks},
		{label: "wantAsset, no exit", want: testAsset},
		{label: "wantAsset, exit seconds 51200", want: testAsset, exit: &seconds},
		{label: "wantAsset, exit seconds 604672", want: testAsset, exit: &long},
		{label: "wantAsset, ratio 1/4", want: testAsset, ratio: [2]uint64{1, 4}},
		{label: "wantBtc, ratio 3/8, exit blocks 4032", offer: testAsset,
			ratio: [2]uint64{3, 8}, exit: &short},
	}

	type vector struct {
		Label     string `json:"label"`
		OfferHex  string `json:"offerHex"`
		Address   string `json:"address"`
		LeafCount int    `json:"leafCount"`
	}
	out := make([]vector, 0, len(cases))
	for _, c := range cases {
		o := &contract.Offer{
			WantAmount: 50_000, WantAsset: c.want, OfferAsset: c.offer,
			RatioNum: c.ratio[0], RatioDen: c.ratio[1],
			MakerPkScript: makerPkScript, MakerPublicKey: maker,
			EmulatorPubkey: emulator, ExitDelay: c.exit,
		}
		vtxoScript, err := o.VtxoScript(server)
		if err != nil {
			panic(err)
		}
		taprootKey, _, err := vtxoScript.TapTree()
		if err != nil {
			panic(err)
		}
		addr := &arklib.Address{HRP: "tark", Signer: server, VtxoTapKey: taprootKey}
		encoded, _ := addr.EncodeV0()
		pkScript, _ := addr.GetPkScript()
		o.SwapPkScript = pkScript
		payload, err := o.Serialize()
		if err != nil {
			panic(err)
		}
		out = append(out, vector{c.label, hex.EncodeToString(payload), encoded,
			len(vtxoScript.Closures)})
	}
	enc := json.NewEncoder(os.Stdout)
	enc.SetIndent("", "  ")
	_ = enc.Encode(out)
}
```

</details>

The tests were checked to be load-bearing by mutation, not just by passing:

| Mutation | Result |
|---|---|
| `exit` moved ahead of `cancel` in the leaf list | 4 vector tests fail |
| ratio records swapped with `offerAsset` in the emitted order | all-optionals vector fails |
| `delay < 512` relaxed to `delay <= 512` | the boundary test fails |
| `exit` dropped from the leaves that mean "returned" | the exit-classification test fails |

The record-order mutation **survived** the first time — no vector carried both an
`offerAsset` and a ratio, so that ordering was untested. The all-optionals vector
exists because of it.

## Test plan

- `pnpm -C packages/swap test` — 600 → 631 passing, no change to the pre-existing
  600.
- `pnpm -C packages/swap run typecheck` — clean.
- `pnpm run build`, `pnpm -C packages/swap run smoke:dist`, `pnpm run test:unit` —
  clean.
- Both golden swap addresses at the top of `offer.test.ts` are unchanged, which is
  the check that no already-published offer's address moved.

## Deliberately not in this PR

- **Partial fills.** The ratio is decoded and carried; nothing interprets it. The
  `fulfill` closure still requires the full `wantAmount`.
- **A distinct `exited` status.** An exit reports as `cancelled` (see below), so
  the record does not say whether the deposit came back cooperatively or
  unilaterally. Adding that distinction means a new `SpendKind` and a new
  persisted status across `TERMINAL`, `RETIRABLE` and every consumer that
  switches on it — worth its own PR, and not needed to keep the swap correct.

## The exit spend had to be classified, not deferred

Adding a third leaf without teaching `classifySpend` about it would have stranded
the swap it belongs to. `classifySpend` matched only `cancel` and `fulfill`, so an
exit spend matched neither and returned `indeterminate` — **deterministically**,
not transiently, unlike the existing `indeterminate` cases which a later poll
clears. From there:

- `spendUpdate` writes nothing for `indeterminate` (`watch.ts:171`), so the swap
  stays `pending` forever;
- `restoreAssetSwaps` adds it to `unresolved` and persists nothing
  (`restore.ts:352`), so every later scan re-queues it;
- and `retireOfferContract` is gated on a `RETIRABLE` status (`watch.ts:187`), so
  the contract is **never retired** — the dead script stays in the subscription,
  the failsafe poll and every sync for the life of the wallet.

So `cancel` and `exit` both report `cancelled`. The vocabulary is what became of
the deposit, not which key moved it: the two differ in who had to agree, not in
outcome. It is also what the wallet's own `isCancelSpend` already infers for an
exit spend, so the two implementations agree rather than diverging.

## Note for solverd

`fetchServerConfig` builds the exit delay as
`RelativeLocktime{Type: LocktimeTypeSecond, Value: uint32(info.UnilateralExitDelay)}`
with no rounding, and both implementations reject seconds that are not a multiple
of 512 (Go: `seconds must be a multiple of 512`; here, bip68's encoder). So an
arkd whose `UnilateralExitDelay` is not 512-granular makes solverd's own
`CreateOffer` fail. Nothing to fix on this side — arkd's own delays are
512-granular (mainnet's is 605184 = 512 × 1182) — but worth knowing.

### CODE REVIEW

File: packages/swap/src/offer.ts [hunk 1]
```diff
@@ -190,28 +261,43 @@ export function encodeOffer(offer: Offer): Uint8Array {
             throw new Error(`${name} must be ${FIELDS[name].width} bytes`);
         }
     }
-    // the wire field is a fixed u64; setBigUint64 would wrap silently past 2^64
-    // (reachable at ~18.45 tokens of an 18-decimal asset) while the covenant
-    // binds the full amount — reject instead of advertising a wrapped amount
-    if (offer.wantAmount < BigInt(0) || offer.wantAmount >> BigInt(64) > BigInt(0)) {
-        throw new Error("wantAmount does not fit the offer wire format (u64)");
+    const ratioNum = setRatio(offer.ratioNum);
+    const ratioDen = setRatio(offer.ratioDen);
```
[the remark this answers is not in the document]
Comment (Kukks) [reply packages/swap/src/offer.ts]: Fixed in `fef4f37`. Confirmed the report: `setRatio(-1n)` returned `undefined` because `-1n > 0n` is false, so a negative was normalized to `unset` and slipped past the `u64` guard entirely — encoding an offer without the ratio the caller asked for.`setRatio` now takes the field name, keeps `0` as the reference's explicit unset, and throws `<name> does not fit the offer wire format (u64)` below zero. Covered by `rejects a negative ratio rather than dropping it as unset`, which exercises a negative in each position.

**Comment by Kukks** [comment]
@coderabbitai review

**Comment by Kukks** [comment]
Thanks — both points were right, and the first is fixed in `0a85513`.

**Zero-value `exitDelay` footgun.** Reproduced before fixing rather than taking it on the report: an explicit `exitDelay: { type: "blocks", value: 0n }` built a **three-leaf** tree and encoded to `0c 0009` + eight zero bytes. So the offer reads as having a unilateral route out — the leaf count says so — while the CSV imposes no wait at all, which is strictly worse than `noExit` because it is harder to notice.

Guarded in `encodeExitDelay` rather than at the `createOffer` call site, so both routes into an offer are covered by one check, and negatives get the same named error instead of the generic `u64` one.

**`indeterminate` forever on an exit spend.** Confirmed at `restore.ts:352` — `unresolved.add(...)` then `continue`, so nothing is persisted and it re-queues on every scan. Left for a follow-up as you suggest, since it needs a new `SpendKind` and a new persisted status. I did amend the comment there: "took neither covenant leaf" was written when there were two leaves, and this PR turns that branch from exceptional into routine, so it should say why.

**One correction to the cross-repo section.** `arkade-os/wallet` (`master`, 179e0b3) *is* a consumer of `createOffer`, `cancelOffer` and `decodeOffer` — `src/providers/assetSwaps.tsx`. I checked it directly: it holds no copy of the offer construction, TLV codec or program JSONs, reads no leaves or `tapLeafScript`, persists `offerHex` so pre-change offers re-derive to the address they were funded at, and passes neither `exitDelay` nor `noExit` — so it picks up the exit closure with no wallet change. Its one local piece, `isCancelSpend`, is inferential and reports an exit spend as a cancel, which is the right economic outcome for a history row.

**Comment by Kukks** [comment]
Follow-up from review: the exit spend needed classifying, not deferring — it would have stranded the swap it belongs to.

`classifySpend` matched only the `cancel` and `fulfill` leaves, so an exit spend matched neither and returned `indeterminate` **deterministically**, unlike the existing `indeterminate` cases which a later poll clears. Traced the consequences rather than assuming they were cosmetic:

- `spendUpdate` writes nothing for `indeterminate` (`watch.ts:171`), so the swap stays `pending` forever.
- `restoreAssetSwaps` adds it to `unresolved` and persists nothing (`restore.ts:352`), so every later scan re-queues it.
- `retireOfferContract` is gated on a `RETIRABLE` status (`watch.ts:187`), so **the contract is never retired** — and `retained` is what drops a script from the subscription, the failsafe poll and every sync. A dead offer script would have stayed in all three for the life of the wallet, growing with every exit.

Fixed in `98d074d`: `cancel` and `exit` both report `cancelled`. The vocabulary is what became of the deposit, not which key moved it — the two differ in who had to agree, not in outcome. It also matches what the wallet's own `isCancelSpend` already infers for an exit spend, so the two implementations agree instead of diverging on the new leaf.

Leaves are collected by name with undefined ones dropped, so an exit-less offer has nothing to match the exit leaf against — a filter that let an undefined leaf through would have matched any spend carrying no tapleaf at all. Both cases are tested, and the classification test fails if `exit` is dropped from the returned-leaf set.

A distinct `exited` status is still deferred: that needs a new `SpendKind` plus a new persisted status across `TERMINAL`, `RETIRABLE` and every consumer that switches on it, and it is not needed to keep the swap correct.

File: packages/swap/src/offer.ts [hunk 2]
```diff
@@ -190,28 +266,50 @@ export function encodeOffer(offer: Offer): Uint8Array {
...
+function encodeExitDelay(exit: RelativeTimelock): Uint8Array {
+    const kind = EXIT_TYPES.indexOf(exit.type);
+    if (kind < 0) throw new Error(`unknown exitDelay locktime type: ${exit.type}`);
+    // a zero delay is an exit in name only: the leaf exists, so the offer reads
+    // as having a unilateral route out, but the CSV imposes no wait at all.
+    // `serverExitDelay` refuses the same value on the default path — this is
+    // what closes it for a caller passing `exitDelay` explicitly.
+    if (exit.value <= BigInt(0)) {
+        throw new Error("exitDelay must be a positive relative locktime");
+    }
```
[the remark this answers is not in the document]
Comment (Kukks) [reply packages/swap/src/offer.ts:342]: Good catch on the ordering — fixed in `bf3e8ef`. Split the finding in two, since I took one half and not the other.

**Validate before registration — agreed, and worse than it looks.** `createOffer` derives the covenant and calls `registerOfferContract` before it encodes, so the checks living in `encodeExitDelay` ran too late. A CSV of zero compiles fine, so an explicit `exitDelay: { value: 0n }` registered a **watched** contract and only then threw. The row is left behind for an offer that never formed — and `promoteOfferContract` has already marked its address outstanding, so `retireOfferContract` will not take it back either. Extracted `assertExitDelay` and run it where the delay is bound.

The server-derived path now goes through the same validator, which the report did not ask for but has the identical defect: a server advertising a delay wider than the wire could carry got as far as `offerVtxoScript`, which failed with bip68's `Expected Number seconds <= 33553920` — naming neither the field nor the way out.

Both are mutation-checked. Reverting to encode-time validation fails `registers nothing for an explicit delay it will refuse to encode` on exactly the leak: `expected [ { type: 'arkade', …(5) } ] to deeply equal []`.

**Reusing the validator in `decodeExitDelay` — declining, deliberately.** The reference (`solverd`, `pkg/swap/contract/offer.go`) emits any `ExitDelay` it is given and its decoder accepts a zero, so rejecting one on the way in would make us unable to read an offer solverd can publish — which is the exact class of bug this PR exists to fix. A zero-delay exit closure also harms only the maker who chose it: it is their own protection being waived, and `cancel` is already untimelocked, so it gives a taker nothing new. Encode stays strict so we never publish one.

I did add the decode test you asked for, pinning that we accept and carry it — `DECODES a zero exit delay it would refuse to emit` — so the asymmetry is recorded as intentional rather than looking like an oversight.

**Comment by Kukks** [comment]
@coderabbitai review

**Comment by Kukks** [comment]
Heads-up for anyone reading the review above: it covers **`8c7b164→0a85513`**, which is three commits behind the head (`2bf5e33`). Both open items it lists were fixed in `98d074d`, before that review was posted.

**"Exit spend → `indeterminate` forever — still open"** — closed in `98d074d`. `classifySpend` now collects both routes that hand the deposit back (`restore.ts:164`, `returned: ["cancel", "exit"]`), and an exit spend reports `cancelled`. The reasoning matches what the review itself argues: the money left correctly, so the question the classifier answers is what became of the deposit, not which key moved it.

The review's own framing of the cost is why it was not deferred — "a permanent subscription/poll/sync entry and a stuck-pending UI record" is a leak that grows by one script per exit, permanently, since the classification is deterministic rather than transient. Mutation-checked: dropping `exit` from the returned-leaf set fails the new test.

**Stale "two leaves" JSDoc** — also gone in `98d074d`. `restore.ts:112-122` now reads "The vocabulary is what became of the deposit, not which key moved it", and states that `exit` reports `cancelled` and why.

Two commits since: `bf3e8ef` moves the exit-delay validation ahead of covenant registration (CodeRabbit finding — a zero delay was registering a watched contract before throwing), and `2bf5e33` pins the derived artifact JSON so the exit closure is readable as data rather than only as TypeScript.

Note that **CodeRabbit has not reviewed `bf3e8ef` or `2bf5e33`** — its check reports green, but the comment says "Review rate limited", and a re-trigger returned "Review rate limited / does not re-review already reviewed commits". So those two commits currently have no bot review at all.

CI is green on `2bf5e33`: all 8 checks, run `33630642218`.
